api/v1beta1/azuremachine_default.go

```diff
@@ -79,10 +84,32 @@ func (s *AzureMachineSpec) SetDataDisksDefaults() {
 }
 
 // SetIdentityDefaults sets the defaults for VM Identity.
-func (s *AzureMachineSpec) SetIdentityDefaults() {
+func (s *AzureMachineSpec) SetIdentityDefaults(subscriptionID string) {
+	// Ensure the deprecated fields and new fields are not populated simultaneously
+	if s.RoleAssignmentName != "" && s.SystemAssignedIdentityRole != nil && s.SystemAssignedIdentityRole.Name != "" {
+		// Both the deprecated and the new fields are both set, return without changes
+		// and reject the request in the validating webhook which runs later.
+		return
+	}
 	if s.Identity == VMIdentitySystemAssigned {
-		if s.RoleAssignmentName == "" {
-			s.RoleAssignmentName = string(uuid.NewUUID())
+		if s.SystemAssignedIdentityRole == nil {
+			s.SystemAssignedIdentityRole = &SystemAssignedIdentityRole{}
+		}
+		if s.SystemAssignedIdentityRole.Name == "" {
+			// Default role name to the role assignment name.
+			s.SystemAssignedIdentityRole.Name = string(uuid.NewUUID())
```

Comment:
super small nit

this could be 

``` 
if s.RoleAssignmentName != "" {
  // Move the existing value from the deprecated RoleAssignmentName field.
  s.SystemAssignedIdentityRole.Name = s.RoleAssignmentName
  s.RoleAssignmentName = ""
} else {
  // Default role name to the role assignment name.
  s.SystemAssignedIdentityRole.Name = string(uuid.NewUUID())
}
``` 

so it's clear that it's one or the other and we avoid running `uuid.NewUUID()` when not needed

Reply:
I tried this before which resulted in the machine controller's ownerRef not being set. I was very confused as to why, and I thought it might have to do with the defaulting webhook being called multiple times, and the Role Assignment Name is overwritten because we aren't checking if `s.SystemAssignedIdentityRole.Name == ""`. Would it make sense to use this structure, but with an extra check for that?

Reply:
yes the way you have it now looks good